tfhe/src/integer/server_key/radix/sub.rs

```diff
@@ -323,4 +324,119 @@ impl ServerKey {
+    /// assert_eq!(expected_result, decrypted_result);
+    /// assert_eq!(expected_overflow, decrypted_overflow);
+    /// ```
+    pub fn unsigned_overflowing_sub(
+        &self,
+        ctxt_left: &RadixCiphertext,
+        ctxt_right: &RadixCiphertext,
+    ) -> (RadixCiphertext, Ciphertext) {
+        let mut tmp_lhs;
+        let mut tmp_rhs;
+
+        let (lhs, rhs) = match (
+            ctxt_left.block_carries_are_empty(),
+            ctxt_right.block_carries_are_empty(),
+        ) {
+            (true, true) => (ctxt_left, ctxt_right),
+            (true, false) => {
+                tmp_rhs = ctxt_right.clone();
+                self.full_propagate_parallelized(&mut tmp_rhs);
+                (ctxt_left, &tmp_rhs)
+            }
+            (false, true) => {
+                tmp_lhs = ctxt_left.clone();
+                self.full_propagate_parallelized(&mut tmp_lhs);
+                (&tmp_lhs, ctxt_right)
+            }
+            (false, false) => {
+                tmp_lhs = ctxt_left.clone();
+                tmp_rhs = ctxt_right.clone();
+                rayon::join(
+                    || self.full_propagate_parallelized(&mut tmp_lhs),
+                    || self.full_propagate_parallelized(&mut tmp_rhs),
+                );
+                (&tmp_lhs, &tmp_rhs)
+            }
+        };
+
+        self.unchecked_unsigned_overflowing_sub(lhs, rhs)
+    }
+
+    pub fn unchecked_unsigned_overflowing_sub(
+        &self,
+        lhs: &RadixCiphertext,
+        rhs: &RadixCiphertext,
+    ) -> (RadixCiphertext, Ciphertext) {
+        assert!(
+            lhs.blocks.len() >= rhs.blocks.len(),
+            "Left hand side must must have a number of blocks greater or equal \
+            to the number of blocks of the right hand side: lhs {} blocks, rhs {} blocks",
+            lhs.blocks.len(),
+            rhs.blocks.len(),
+        );
+        let modulus = self.key.message_modulus.0 as u64;
+
+        // If the block does not have a carry after the subtraction, it means it needs to
+        // borrow from the next block
+        let compute_borrow_lut = self
+            .key
+            .generate_lookup_table(|x| if x < modulus { 1 } else { 0 });
+
+        let mut borrow = self.key.create_trivial(0);
+        let mut new_blocks = Vec::with_capacity(lhs.blocks.len());
+        for (lhs_b, rhs_b) in lhs.blocks.iter().zip(rhs.blocks.iter()) {
+            let mut result_block = self.key.unchecked_sub(lhs_b, rhs_b);
+            // Here unchecked_sub_assign does not give correct result, we don't want
+            // the correcting term to be used
+            // -> This is ok as the value returned by unchecked_sub is in range 1..(message_mod * 2)
+            crate::core_crypto::algorithms::lwe_ciphertext_sub_assign(
+                &mut result_block.ct,
+                &borrow.ct,
+            );
+            let (msg, new_borrow) = rayon::join(
```

Comment:
taking a simple case where lhs is a single block and rhs as well where lhs = 0 and rhs = 2 I seem to not find the right result

the unchecked sub produces the modular opposite so -2 mod msg_mod which is greater than msg_mod

the borrow is zero, so the next sub still has that -2 mod msg_mod

the apply lookup table will return a message encrypting a zero right ?

am I missing something ?

Reply:
Lhs: [0]
Rhs: [2]

Lhs - Rhs: [0 - 2 + msg_ms] = [0 - 2 + 4] = [2]
2 < msg_mod -> output_borrow 1.

So at the end we have (0 - 2) % 4 = 2 with an output borrow of one, meaning overflow occured

Reply:
thanks !